src/main/java/duke/task/TaskList.java

```diff
@@ -0,0 +1,151 @@
+package duke.task;
+
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import duke.Ui;
+import duke.command.CommandKeyword;
+import duke.exception.DukeException;
+import duke.exception.IncompleteTaskDescriptionException;
+import duke.exception.InvalidCommandException;
+
+/**
+ * Contains the task list. It has operations to change the tasks in the list.
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private Ui ui;
+
+    /**
+     * Constructor for TaskList.
+     *
+     * @param tasksInfo Arraylist of tasks loaded from storage.
+     * @param ui An Ui instance to deal with interactions with the user.
+     */
+    public TaskList(ArrayList<String> tasksInfo, Ui ui) {
+        List<Task> list = tasksInfo.stream().map(Task::convertToTask).collect(Collectors.toList());
+        this.tasks = new ArrayList<>(list);
+        this.ui = ui;
+    }
+
+    /**
+     * Display the list of tasks to the user.
+     */
+    public void display() {
+        ui.showTasks(this.tasks);
+    }
+
+    /**
+     * Mark the task at the given index.
+     *
+     * @param index The index of the arraylist that contains the task.
+     * @return The updated task list after the task is marked.
+     */
+    public ArrayList<Task> markTask(int index) {
+        try {
+            Task task = this.tasks.get(index);
+            task.markAsDone();
+            this.ui.showMarkedTask(task);
+            return this.tasks;
+        } catch (IndexOutOfBoundsException e) {
+            this.ui.showMarkedTask(null);
+            return null;
+        }
+    }
+
+    /**
+     * Add the task to the task list.
+     *
+     * @param description The description of the task.
+     * @param keyword The command keyword to identify the type of task to be added.
+     * @return The updated task list after the task is added.
+     * @throws DukeException If the task fails to be added to the list.
+     */
+    public ArrayList<Task> addTask(String description, CommandKeyword keyword) throws DukeException {
+        Task task = null;
```

Comment:
Would it be better if you do the declaration and initialisation together in the respective case blocks, and doing `tasks.add(task)` in the case block as well? I think that this has the benefit of ensuring that the task will never be `null`, especially if the logic is changed in some way in future.